onnxruntime/core/session/onnxruntime_c_api.cc

```diff
@@ -630,6 +632,421 @@ ORT_API_STATUS_IMPL(OrtSessionGetOutputName, _In_ const OrtSession* sess, size_t
 }
 
+///////////////////////////////////////////////////////////////////////////
+// Code to handle non-tensor types
+// OrtGetNumValues
+// OrtGetVaue
+// OrtCreateValue
+///////////////////////////////////////////////////////////////////////////
+const int NUM_MAP_INDICES = 2;
+
+////////////////////
+// OrtGetNumValues
+template <typename T>
+OrtStatus* OrtGetNumSequenceElements(const MLValue* p_ml_value, size_t* out) {
+  auto& data = p_ml_value->Get<T>();
+  *out = data.size();
+  return nullptr;
+}
+
+static OrtStatus* OrtGetNumValuesImpl(const OrtValue* value, size_t* out) {
+  auto value_type = OrtGetValueType(value);
+  if (value_type == ONNX_TYPE_MAP) {
+    *out = NUM_MAP_INDICES;
+    return nullptr;
+  } else if (value_type == ONNX_TYPE_SEQUENCE) {
+    auto v = reinterpret_cast<const MLValue*>(value);
+    auto type = v->Type();
+    // Note: keep these in sync with the registered types in data_types.h
+    if (type == DataTypeImpl::GetType<VectorString>()) {
+      return OrtGetNumSequenceElements<VectorString>(v, out);
+    } else if (type == DataTypeImpl::GetType<VectorInt64>()) {
+      return OrtGetNumSequenceElements<VectorInt64>(v, out);
+    } else if (type == DataTypeImpl::GetType<VectorFloat>()) {
+      return OrtGetNumSequenceElements<VectorFloat>(v, out);
+    } else if (type == DataTypeImpl::GetType<VectorDouble>()) {
+      return OrtGetNumSequenceElements<VectorDouble>(v, out);
+    } else if (type == DataTypeImpl::GetType<VectorMapStringToFloat>()) {
+      return OrtGetNumSequenceElements<VectorMapStringToFloat>(v, out);
+    } else if (type == DataTypeImpl::GetType<VectorMapInt64ToFloat>()) {
+      return OrtGetNumSequenceElements<VectorMapInt64ToFloat>(v, out);
+    } else {
+      return OrtCreateStatus(ORT_FAIL, "Input is not of one of the supported sequence types.");
+    }
+  } else {
+    return OrtCreateStatus(ORT_FAIL, "Input is not of type sequence or map.");
+  }
+}
+
+ORT_API_STATUS_IMPL(OrtGetNumValues, const OrtValue* value, size_t* out) {
+  API_IMPL_BEGIN
+  return OrtGetNumValuesImpl(value, out);
+  API_IMPL_END
+}
+
+///////////////////
+// OrtGetValue
+template <typename T>
+static OrtStatus* OrtGetValueImplSeqOfMap(const MLValue* p_ml_value, int index,
+                                          OrtValue** out) {
+  using TKey = typename T::value_type::key_type;
+  using TVal = typename T::value_type::mapped_type;
+  using MapType = std::map<TKey, TVal>;
+  auto& data_vec = p_ml_value->Get<T>();
+  auto& data_elem = data_vec.at(index);
+  auto copy_data_elem = std::make_unique<MapType>(data_elem);
+  std::unique_ptr<MLValue> value = std::make_unique<MLValue>();
+  value->Init(copy_data_elem.release(),
+              DataTypeImpl::GetType<MapType>(),
+              DataTypeImpl::GetType<MapType>()->GetDeleteFunc());
+  *out = reinterpret_cast<OrtValue*>(value.release());
+  return nullptr;
+}
+
+template <typename T>
+static ONNXTensorElementDataType GetONNXTensorElementDataType() {
```

Comment:
Is it possible to specialize this function by type? Or do othe comparisons optimize away completely in release? Too bad we can't use if constexpr...

Reply:
sure.